src/kernel/heap.zig

```diff
@@ -84,16 +90,23 @@ pub const Heap = struct {
     /// See std/mem.zig for documentation. This function should only be called by the Allocator interface.
     fn shrink(allocator: *Allocator, old_mem: []u8, old_alignment: u29, new_size: usize, new_alignment: u29) []u8 {
         var heap = @fieldParentPtr(Heap, "allocator", allocator);
-        if (new_size == 0) {
+        if (new_size != old_mem.len) {
             // Freeing will error if the pointer was never allocated in the first place, but the Allocator API doesn't allow errors from shrink so use unreachable
             // It's not nice but is the only thing that can be done. Besides, if the unreachable is ever triggered then a double-free bug has been found
             heap.free(@ptrToInt(&old_mem[0]), old_mem.len) catch unreachable;
+            if (new_size != 0) {
```

Comment:
Probs don't need to add this now, but maybe add a issue, but to make this more efficient, could just free the memory that is to be shrinked and not all the memory then allocate again. Like `heap.free(@ptrToInt(&old_mem[new_size]), old_mem.len - new_size) catch unreachable;` If this easy to do, then could do this.

Reply:
That is a nice idea, but I don't think it would work all the time because of how the buddy system is structured. If the start of the memory that is no longer needed is in the same block as the memory that is still needed all of it will be freed. For example if `old_mem` starts at address 0, has length 32 and the min block size is 16 then a shrink to length 10 would try to free 11..32, but because of how the structure works this would end up freeing the whole allocation from 0 to 32, which we wouldn't want. If we find a way to do it more efficiently than freeing and re-allocating that would work in all cases I'm up for it :)

Reply:
Ah yes, so could have `new_size` be aligned up to the nearest block size; `std.mem.alignForward(new_size, block_size)`

Reply:
I think that would still have the same issue

Reply:
Ok, can create an issue maybe to say that there could be some optimisation here.

Reply:
Good idea :)